deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/UrlTrie.kt

```diff
@@ -0,0 +1,143 @@
+package com.airbnb.deeplinkdispatch
+
+import com.airbnb.deeplinkdispatch.base.MatchIndex
+import com.airbnb.deeplinkdispatch.base.MatchIndex.*
+import java.io.OutputStream
+import java.nio.charset.Charset
+import kotlin.text.Charsets.UTF_8
+
+data class UriMatch(val uri: DeepLinkUri, val matchId: Int)
+
+@kotlin.ExperimentalUnsignedTypes
+open class TrieNode(open val id: String = "", val type: UByte, open val placeholder: Boolean = false) {
+
+    val children = mutableSetOf<TrieNode>()
+    var match: UriMatch? = null
+        set(value) {
+            if (match != null) error("Ambiguous URI. Same match for two URIs ($match vs $value)") else field = value
+        }
+
+    fun addNode(node: TrieNode): TrieNode {
+        return if (children.add(node)) node else children.find { it.equals(node) }!!
```

Comment:
`!!` is a bit of a red flag. can use `first` instead if that is the behavior you want

also, `==` instead of `.equals`